core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1962,6 +1963,40 @@ protected Type visitCurrentPath(CurrentPath node, StackableAstVisitorContext<Con
             return setExpressionType(node, VARCHAR);
         }
 
+        @Override
+        protected Type visitTrim(Trim node, StackableAstVisitorContext<Context> context)
+        {
+            ImmutableList.Builder<Type> argumentTypes = ImmutableList.builder();
+
+            argumentTypes.add(process(node.getTrimSource(), context));
+            node.getTrimCharacter().ifPresent(trimChar -> argumentTypes.add(process(trimChar, context)));
+            List<Type> actualTypes = argumentTypes.build();
+
+            String functionName = node.getSpecification().getFunctionName();
+            ResolvedFunction function = plannerContext.getMetadata().resolveFunction(session, QualifiedName.of(functionName), fromTypes(actualTypes));
+
+            List<Type> expectedTypes = function.getSignature().getArgumentTypes();
+            if (expectedTypes.size() != 1 && expectedTypes.size() != 2) {
+                throw semanticException(INVALID_PARAMETER_USAGE, node, "Trim function takes one or two parameters");
+            }
```

Comment:
If the resolved signature doesn't match the actual argument number, that would be an internal error, not a user error.
Instead of this check, you can do just a simple checkState:
`checkState(expectedTypes.size() == actualTypes.size(), "wrong argument number in the resolved signature")`